byzcoin/contract_deferred.go

```diff
@@ -0,0 +1,404 @@
+	//   - Handle the "addProof" invocation
+	//   - Handle the "execProposedTx" invocation
+	//
+	// Invoke:addProof should have the following input argument:
+	//   - identity darc.Identity
+	//   - signature []byte
+	//   - index uint32 (index of the instruction wrt the transaction)
+	err2 := c.checkInvoke(rst, inst.Invoke)
+	if err2 != nil {
+		return nil, nil, errors.New("checks of invoke failed: " + err2.Error())
+	}
+
+	cout = coins
+
+	// Find the darcID for this instance.
+	var darcID darc.ID
+
+	_, _, _, darcID, err = rst.GetValues(inst.InstanceID.Slice())
+	if err != nil {
+		return
+	}
+
+	switch inst.Invoke.Command {
+	case "addProof":
+		// This invocation appends the identity and the corresponding signature,
+		// which is based on the stored instruction hash (in instructionHashes)
+
+		// Get the given index
+		indexBuf := inst.Invoke.Args.Search("index")
+		if indexBuf == nil {
+			return nil, nil, errors.New("index args is nil")
+		}
+		index := binary.LittleEndian.Uint32(indexBuf)
+
+		// Check if the index is in range
+		numInstruction := len(c.DeferredData.ProposedTransaction.Instructions)
+		if index >= uint32(numInstruction) {
+			return nil, nil, fmt.Errorf("index is out of range (%d >= %d)", index, numInstruction)
+		}
+
+		// Get the given Identity
+		identityBuf := inst.Invoke.Args.Search("identity")
+		if identityBuf == nil {
+			return nil, nil, errors.New("identity args is nil")
+		}
+		identity := darc.Identity{}
+		err = protobuf.Decode(identityBuf, &identity)
+		if err != nil {
+			return nil, nil, errors.New("couldn't decode Identity")
+		}
+
+		// Get the given signature
+		signature := inst.Invoke.Args.Search("signature")
+		if signature == nil {
+			return nil, nil, errors.New("signature args is nil")
+		}
+		// Update the contract's data with the given signature and identity
+		c.DeferredData.ProposedTransaction.Instructions[index].SignerIdentities = append(c.DeferredData.ProposedTransaction.Instructions[index].SignerIdentities, identity)
+		c.DeferredData.ProposedTransaction.Instructions[index].Signatures = append(c.DeferredData.ProposedTransaction.Instructions[index].Signatures, signature)
+		// Save and send the modifications
+		cosiDataBuf, err2 := protobuf.Encode(&c.DeferredData)
+		if err2 != nil {
+			return nil, nil, errors.New("couldn't encode DeferredData")
+		}
+		sc = append(sc, NewStateChange(Update, inst.InstanceID,
+			ContractDeferredID, cosiDataBuf, darcID))
+		return
+	case "execProposedTx":
+		// This invocation tries to execute the transaction stored with the
+		// "Spawn" invocation. If it is successful, this invocation fills the
+		// "ExecResult" field of the "deferredData" struct.
+
+		// In the following we are creating a new StagingStateTrie from the
+		// readonly state try by copying the data.
+		nonce, err2 := rst.GetNonce()
+		if err2 != nil {
+			return nil, nil, errors.New("couldn't get the nonce: " + err2.Error())
+		}
+		sst, err2 := newMemStagingStateTrie(nonce)
```

Comment:
the places with `err2` can be changed to use named return values

Reply:
In the case of assigning multiple variable where one of them must be initialized, I can't use `err`. For example, with `sst, err := newMemStagingStateTrie(nonce)` I get "`err` is shadowed during return", because I used the short variable declaration. 

In this case I would need to to the following: 

```go
		var sst *stagingStateTrie
		sst, err = newMemStagingStateTrie(nonce)
		if err != nil {
                        err = errors.New("Failed to created stagingStateTrie: " + err2.Error())
			return
		}
``` 

Instead of

```go
		sst, err2 := newMemStagingStateTrie(nonce)
		if err2 != nil {
			return nil, nil, errors.New("Failed to created stagingStateTrie: " + err2.Error())
		}
```

Which solution should I pick?